packages/core/src/sagas/startDictateAfterSpeakActivitySaga.js

```diff
@@ -1,38 +1,43 @@
 import {
-  cancel,
-  fork,
   put,
   select,
-  take
+  takeEvery
 } from 'redux-saga/effects';
 
 import whileConnected from './effects/whileConnected';
+import whileSpeakIncomingActivity from './effects/whileSpeakIncomingActivity';
 
 import { MARK_ACTIVITY } from '../actions/markActivity';
-import { START_SPEAKING_ACTIVITY } from '../actions/startSpeakingActivity';
-import { STOP_SPEAKING_ACTIVITY } from '../actions/stopSpeakingActivity';
 import startDictate from '../actions/startDictate';
 
 export default function* () {
-  yield whileConnected(function* (_, userID) {
-    for (;;) {
-      yield take(START_SPEAKING_ACTIVITY);
-
-      const task = yield fork(startDictateAfterSpeakActivitySaga, userID);
-
-      yield take(STOP_SPEAKING_ACTIVITY);
-      yield cancel(task);
-    }
+  yield whileConnected(function* () {
+    yield whileSpeakIncomingActivity(startDictateAfterSpeakActivitySaga);
   });
 }
 
 function* startDictateAfterSpeakActivitySaga() {
-  for (;;) {
-    const { payload: { activityID } } = yield take(({ payload, type }) => type === MARK_ACTIVITY && payload.name === 'speak' && payload.value === false);
-    const activities = yield select(({ activities }) => activities);
+  yield takeEvery(
+    ({ payload, type }) => (
+      type === MARK_ACTIVITY
+      && payload.name === 'speak'
+      && payload.value === false
+    ),
+    function* ({ payload: { activityID } }) {
+      const { activities } = yield select();
```

Comment:
Like an earlier comment, I'm favoring a defined selector that can represent the filter below and made available for reuse.

Is something like this possible?
```javascript
export const findActivity = state = id => (state.activities.filter(activity => (activity.id === id)))
```

```javascript
const activity = select(findActivity(id))
```

Reply:
I agree with the previous one. But this one doesn't looks like have a lots of benefits to me.

For example, `findActivityByID` is a function that find an activity in a set by ID, and given ID is unique in the set.

The original code looks like:

```js
const spokenActivity = activities.find(({ id }) => id === activityID);
```

The new code would looks like:

```js
const spokenActivity = findActivityByID(activityID);
```

And the utility/selector function `findActivityByID.js` will looks like:

```js
export default function (activityID) {
  activities.find(({ id }) => id === activityID);
}
```

The benefit of finding one activity is not significant in this case, and also given that the schema is locked down, we are very unlikely to change how we handle activity ID.

Another potential possible move is a `activityHasID` predicate.

```js
const spokenActivity = activities.find(activity => activityHasID(activity, activityID));
```

And the predicate function `activityHasID.js` would looks like:

```js
export default function (activity, activityID) {
  return activity && activity.id === activityID;
}
```

This coding style is more targeted for adopting changing schema.

What do you think?

Reply:
Resolving this one because of dupe of another comment. Both talks about selector.